### PR TITLE
feat(theme-classic): store selected tab in query string.

### LINKED ISSUE #7008: Open respective tab item with URL query string
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

Can multi-language support code blocks, support anchors which allow the anchored code block to open?

https://docusaurus-five-rho.vercel.app/docs/markdown-features/code-blocks#multi-language-support-code-block

For example, I would like to be able to offer a user a link, either via the docusaurus side bar, or directly, that allows them to jump to a language specific block.

EG clicking Java https://docusaurus-five-rho.vercel.app/docs/markdown-features/code-blocks#java would navigate to that part of the page and make that tab visible.

Screenshot 2022-03-25 at 20 55 52

- The anchor link from the side bar will only work if the anchor is visible
- Hotlinking via the full url plus anchor if in a code block, will not move to the code block even if the anchor is in a visible tab.

This might just be a limitation of my HTML scripting foo, as is possible, if so, please show me the way!

### Has this been requested on Canny?

_No response_

### Motivation

we support an open source project which has a significant feature set across multiple language implementations.

I would like to be able to connect this up with [syncing tab choices ](https://docusaurus-five-rho.vercel.app/docs/markdown-features/tabs#syncing-tab-choices) as it would enable readers to ensure they only see information that is relevant to them, in their preferred programming language.

### API design

_No response_

### Have you tried building it?

Yes, 
See example from docusaurus master, only addition is https://github.com/YOU54F/docusaurus/commit/0e297b966a34e68c4d612a42b1ad3f634d923395 to insert anchors on each of the code block tabs.

Deployed here via vercel however the headers aren't showing 

https://docusaurus-you54f.vercel.app/docs/markdown-features/code-blocks

You can clone my master branch `g clone git@github.com:YOU54F/docusaurus.git`

`npm install && cd website && npm start`

Headers added and visible locally

Screenshot 2022-03-25 at 21 47 59

clicking from TOC will anchor to Javascript tab

Screenshot 2022-03-25 at 21 48 05

If java tab is pre selected and therefore visible

Screenshot 2022-03-25 at 21 48 14

clicking from TOC will anchor to Java tab

Screenshot 2022-03-25 at 21 48 20

Hot linking to any of the code block anchors
 http://localhost:3000/docs/markdown-features/code-blocks#javascript even the visible javascript one will take the user to the top of the page. 
 

### Self-service

- [X] I'd be willing to contribute this feature to Docusaurus myself.

### PR DESCRIPTION
**Pull request by mturoci**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Implementation description

* Selection stored according to [this comment](https://github.com/facebook/docusaurus/issues/7008#issuecomment-1152540679) - serialized JSON with group ids as keys and `__noGroup__` constant key for tabs that do not belong to any group. Downsides:
   * Ugly links - http://localhost:3000/docs/markdown-features/tabs?tabs=%7B%22__noGroup__%22%3A%22banana%22%2C%22operating-systems%22%3A%22win%22%7D
   * In case of multiple tabs that do not belong to any group and have different tab item keys, this solution would only work for the last selected tab.
* Selected key priority: search query > local storage > defaultValue prop > 1st tab with default specified > 1st tab.
* @slorber can you please provide further explanation for [this comment](https://github.com/facebook/docusaurus/issues/7008#issuecomment-1157484387) ? I identified the file that needs the changes as [this](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-classic/src/index.ts), but it's not clear to me what is actually needed to be done. From my understanding, when you hydrate React components and build static HTML off it, you cannot know the search URL query params as they are only available during the actual request or am I missing something?

Let's use this PR as starting point for further discussion around UX/implementation.
## Motivation

Allows for sharing tabs selection via URL - better UX.

## Test Plan

Manual

### Test links

- https://deploy-preview-8225--docusaurus-2.netlify.app/docs/markdown-features/tabs#query-string
- https://deploy-preview-8225--docusaurus-2.netlify.app/docs/markdown-features/tabs?current-os=ios#query-string
- https://deploy-preview-8225--docusaurus-2.netlify.app/docs/markdown-features/tabs?current-os=android#query-string

## Related issues/PRs

Closes #7008

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:88]: can you explain the intent here? Not sure to understand

**Review by slorber — CHANGES_REQUESTED** [review]
Thanks

Not good enough to merge as is.

Sync between groups is now broken: https://deploy-preview-8225--docusaurus-2.netlify.app/docs/markdown-features/tabs/#syncing-tab-choices

I'd rather not serialize as JSON.

I'd also prefer if we opt-in for querystring serialization instead of making it a default

What I'd do:
- provide `queryString: string | boolean` prop
- false by default => no querystring
- true => use groupId as querystring param
- string => use value as querystring param

**Comment by slorber** [comment]
For the hydration problem, I don't know exactly how we can solve it, but if would be really preferably if we didn't have that "hydration tab change" when React loads:

CleanShot 2022-10-20 at 19 47 23

This can be particularly disturbing in case tabs have different heights. If an anchor is provided in addition to qs tabs, it can also mess-up with the scroll position.

Now this is likely not easy to solve so I won't block the PR if it can't be fixed, it can be handled later.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:81]: Very similar to the [prev check](https://github.com/facebook/docusaurus/blob/e41133245ca4e6cee390da74ccfe305ccb2ddb25/packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx#L79). However, this new one only throws in case of user-error. E.g. available tab keys are `ke1`, `key2`, `key3`, but user supplies `key4`. The goal is to notify user it's their fault that the tabs are not working properly.

The prev implementation would throw even if the error was on the Docusaurus (implementation) side as it checked the derived value instead of user supplied one.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: Good point, replaced with react router instead.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: Done.

**Comment by mturoci** [comment]
> preferably if we didn't have that "hydration tab change" when React loads

Yes, the desired outcome is clear. The implementation is the question here since docusaurus uses SSG, not SSR. Curious about the solution though.

@slorber I have added the desired react router APIs, but unfortunately these break unit tests as they are not renderable on the server. I wasn't able to find similar scenario (react router APIs within unit tests) throughout the codebase. Could you please give me some pointers? Thanks!

File: packages/docusaurus-theme-classic/src/theme/Tabs/__tests__/index.test.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/__tests__/index.test.tsx:17]: There's a component to use in tests:

```jsx
<StaticRouter location={{ pathname: "/bubblegum" }}>
  Content using useLocation here
</StaticRouter>
```

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:145]: I guess we could throw if queryString is true but groupId is falsy?

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:148]: I would rather use replace than push. push creates new history entries ie you'll move back to previous tab when navigating. Maybe we could make this configurable later but replace looks like a better default for this specific UX. 

I can't test it because no more dogfood way => we need some, please add some docs about that new attribute, that should make it easier to review/test the PR through the deploy preview

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:101]: that's duplicated logic. I'd like such logic to be extracted in a custom hook or something to avoid the duplication

What about:

```jsx
const tabQueryString = useTabQueryString()

tabQueryString.get();
tabQueryString.set(val);
```

You could also encapsulate the usage of router/history hooks there

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:97]: I'm not even sure we should consider the querystring as a default value

I believe the default value should rather be static (otherwise it could create weird react hydration issues because of a mismatch between server/client)

The actual state value could be "updated" to reflect the querystring in a `useEffect` after mount. I think the `useState` initial value should not be affected by a dynamic param such as the query string.

It may not be a problem today but we'll likely get a warning after upgrading to React 18 if we don't do this.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: We have a dogfood section in our doc where you can add such test: that would be nice to add this test actually to check if it's well-supported.

I think it's not because you don't have any `useEffect(() => {},[qsValue])` that will resync the querystring value to state after mount

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: I'd rather move any code to a custom hook as long as it's possible to create a good abstraction. Here it's the case for me.

https://kyleshevlin.com/use-encapsulation


> which may be overkill compared to just going with a couple lines of code

I don't see how a function call is overkill here. We could implement a whole app in a single file / function if we wanted, and it's not really overkill to split it into well-encapsulated concerns so that you are able to browse your app at different levels of abstractions

See also Clean code rule - Single level of abstraction

**Review by slorber — CHANGES_REQUESTED** [review]
Thanks, looks better but not yet good enough 🤪

**Comment by slorber** [comment]
> Yes, the desired outcome is clear. The implementation is the question here since docusaurus uses SSG, not SSR. Curious about the solution though.

Unfortunately that's exactly why it's not easy. There are tricks to work around this in SSG (see dark mode or ThemedImage). Don't worry, we'll eventually figure out later in another PR.

File: packages/docusaurus-theme-classic/src/theme/Tabs/__tests__/index.test.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/__tests__/index.test.tsx:17]: Great, thanks!

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:151]: Good point, done.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:107]: I prefer to sacrifice a little bit of duplication (single line in this case) in order to avoid unnecessary abstractions. However, I noticed you are in the other camp which I respect :) Refactored.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: `replace` used instead of `push`

Also added docs which should be enough for dogfooding, but let me know if you would like me to add more.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:129]: Refactored into `useEffect` as requested.

**Comment by mturoci** [comment]
> looks better but not yet good enough 🤪

No worries, it's expected to be an iterative process so that we are on the same page as for the required implementation direction :)

> I don't see how a function call is overkill here.

Yea, I see we are from different camps. Clean code is very subjective and although I don't agree with your opinions, I fully respect them and since you are the maintainer here, you have the upper hand ofc :D Refactored into more abstractions as requested. Let me know if that's enough or you would like me to split the custom hook into a separate file as well.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx
Comment (mturoci) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx]: > you can add such test
Not sure what exactly you mean, I have added a docs section to Tabs page which should be enough for dogfooding, but if not, just let me know.

> you don't have any useEffect(() => {},[qsValue]) that will resync the querystring value to state after mount

This is added now.

**Comment by jnikhila** [comment]
Hey! We are keen to use this feature for our documentation. Could you please confirm the release date?

**Comment by slorber** [comment]
Did some refactoring, fixes and docs improvements => should be ready to review. 

Some things to review more carefully:
- clicking a tab should not lead to scroll position changes (it did due to core scrollAfterNavigation())
- how groupId and queryString work together => it makes sense to me the url qs choice overrides the localStorage choice

Please test the preview and let me know if you see anything weird

**Review by slorber — APPROVED** [review]

**Comment by slorber** [comment]
Looks like it works as I want it to 👍

**Comment by zeluspudding** [comment]
May be a silly question - how can one access this feature update before v2.3 becomes live?

**Pull request merged**
